### DISCUSSION TITLE
Detecting collision for a wheeled robot.

### DISCUSSION THREAD

**Discussion by TechnicBRICKs** [pybricks/support#244] [Q&A]
Hi,

Generally I'm trying to detect collision events for a wheeled robot. It can be a fixed obstacle or in the limit another moving car, but lets keep it simple to start.

I've planned to use the IMU from TechnicHub to detect a deceleration, but the fact is that even considerable speed and high sampling rates, the acceleration readings doesn't seem to stand out from the IMU readings noise level.

Another possibility would be to try detecting the wheels slow down. However my experience tells me me that even in front of an obstacle, a relative light weight car keeps its wheels running as if it was moving freely...
This regardless if it is a varnished parquet, ceramic tiles or ABS like the large 8x16 tiles.
I may try playing with stall_tolerances method in Control Class, but I've some doubts this could successfully solve the problem.   
Any ideas you'd like to suggest?

Below the code I'm using to log the IMU readings in the console. For the case of my prototype, the significant readings should appear mostly on Y axis.

```python
while True:
    drive_base.drive(300,0)
    print(hub.imu.acceleration(axis=Axis.X), hub.imu.acceleration(axis=Axis.Y), hub.imu.acceleration(axis=Axis.Z))
    wait (5)
```

I've tried with larger wait times as it was my intention, trying even to remove the wait to increase the sampling to the maximum so I won't miss the collision event, without success.

BTW, I've also experienced another problem.
Sometimes when I stop the program, drivebase motors keep running forever. I may eventually stop them with a long press on hub's green button, but sometimes really need to remove the batteries so that the program can be stopped.

**Comment by TechnicBRICKs** [comment]
Have been trying this code, without effect.

```python
# Main
print("Start: BCar")
light.on(brightness = light_intensity)
    
motor_left.control.stall_tolerances(car_speed, stall_time)
motor_right.control.stall_tolerances(car_speed, stall_time)
    
while True:
    drive_base.drive(car_speed, car_fwd_turn_rate)
    
    if motor_left.control.stalled() or motor_right.control.stalled():
        drive_base.stop()
        drive_base.straight(car_reverse)  
```

The parameters may need fine tuning, but not even if I hold one or the two motors by and, I can trigger the stalled condition.

Not sure if I can access motor_XX.control.stalled() while the drive_base is running.
According to the documentation "You cannot use these motors individually while the DriveBase is active."
Although if not possible this way, how to use it with drive_base?
Apparently can't apply .control.stall_tolerances() directly to the drive_base, hence my approach.

**Comment by TechnicBRICKs** [comment]
Still regarding the usage of IMU to detect a collision, still wondering if this would be feasible...

Prime and Inventor hubs support tap detection with their native firmware tap detection, which work fine with relative weak taps. Hence it should a matter to find out how this is implemented.
Still need to check if a 300mm/s car collision is enough to trigger a tap detection event, or if the velocity needs to be higher and how much.

On this regard I've also another question. Which are the acceleration units returned by IMU sensor with Pybricks?
- LEGO firmware on SPIKE and Inventor hubs return around 1 on Z-axis at rest, hence I believe these are G units.
- TechnicHub with Pybricks returns around 10 on Z-axis at rest, hence I believe these are SI units (m/s2).
 
1G ~= 9,81 m/s2

**Reply by laurensvalk** [reply]
> how this is implemented.

Yes, the sensor can be configured with all kinds of filters. For now we only return the raw values (scaled to appropriate units, see below).

It would be good to have some discussion and a list of use cases, so we can decide on the right settings. So one use case is impact detection like yours. Let's try to make a list of all kinds of use cases.

> Which are the acceleration units returned by IMU sensor with Pybricks?

![image](https://user-images.githubusercontent.com/12326241/106496403-23f83e00-64bd-11eb-881e-0f0e9ba0dbf3.png)

**Reply by TechnicBRICKs** [reply]
My bad!... I read the Pybricks documentation, and later when posting here forgot the units were there. Perhaps suggested by the fact that the same information does not exist regarding LEGO firmware...

As you refer to scale the raw values to appropriate units, you might know which are the IMU raw values units? **G**s I guess!?

As for the use cases tap is indeed a useful one, but I can think about a few other:
- Double tap, triple tap
- Directional tap (+/-x, +/-y, +/-z) 
- Free fall detection, like in origina SPIKE software
- I've made a model where I use SPIKE display to show X/Y acceleration in very basic polar graph, and compare next to the readings in my car dashboard (might publish a video of it some day...).
   Hence the raw reading are useful. In these case I display in **G**s, but that's a conversion everyone can do once the readings units are known.

**Reply by laurensvalk** [reply]
> As you refer to scale the raw values to appropriate units, you might know which are the IMU raw values units? Gs I guess!?

That's configurable also. You can select higher range (higher maximum detection) at the cost of less resolution. Ideally we find the right balance for most applications so we don't have to make commands to set each and every option. 

Directional tap would be cool, I'd have to check if the sensor can do that.

**Comment by TechnicBRICKs** [comment]
Still tinkering around this... 😊

IMU readings are still my preferred option, if viable to make them work reliably or close to that.
Hence I decided to make some additional measurements and graphs to check whether there are readings that standout during a vehicle collision event, or not. My previous analysis were made through a quick inspection at the printouts to the Python console, which is definitely not a reliable way to look at it...

In an attempt to get more insights, I've attached a SPIKE hub to my drivebase, which is powered by the Technic hub, and logged the respective IMU readings through the LEGO SPIKE Prime app (hope you don't mind that I show them here) to compare both logs and eventually take some insights to tweak Pybricks IMU parameters setting.

Some of my initial observations are:
- Car collision at tested speeds were not enough to trigger a 'tap' event, as detected by the SPIKE LEGO app. Although we can often find IMU acceleration spikes in the expected axis (Y), but sometimes also at X, and more often and surprisingly more prominent at Z.
- Sometimes visible spikes do not show up at all.
- And sometimes it would be difficult to identify a possible collision, because there are several reading spikes in one axis or in different axis at different points in time...
- If a collision event needs to be detected in real-time, many existing dataset processing techniques could not be used effectively as we can't wait for future data to arrive for a complete dataset with previous and post spike data points. Hence some expedite algorithm might be considered.

As a preliminary conclusion, the IMU with current Pybricks implementation seems to have some potential to successfully identify collisions, with some false positives and some false negatives along the way.
If complemented with a "slippery" wheels slow down detection, I believe it could be a viable solution to reduce the false negatives.
  
Below some samples and comments. Please note that Technic hub data displayed in Excel charts were measured with Pybricks, and correspondent SPIKE hub data is measured with LEGO firmware and SPIKE app.
Added vertical orange lines to mark the presumable moment of impact, where possible.


**1)** Well behaved and consistent data.
![1A](https://user-images.githubusercontent.com/67207661/106677798-51beaf00-65b1-11eb-9a90-032c19b03039.png)
![1B](https://user-images.githubusercontent.com/67207661/106677803-54b99f80-65b1-11eb-9e5f-ec706a40f1db.png)


**2)** Another well behaved example.
![2A](https://user-images.githubusercontent.com/67207661/106677813-58e5bd00-65b1-11eb-9719-028682cacc54.png)
![2B](https://user-images.githubusercontent.com/67207661/106677824-5be0ad80-65b1-11eb-912c-7e0cf728deb5.png)


**3)** Consistent data, although some unexpected peaks at Y axis. Should get confirmation at another axis, within a a defined short time interval.
![3A](https://user-images.githubusercontent.com/67207661/106677894-8468a780-65b1-11eb-8044-ea92fe06a920.png)
![3B](https://user-images.githubusercontent.com/67207661/106677918-964a4a80-65b1-11eb-89e2-fbeeabd65b06.png)


**4)** Well behaved example with X, Y and Z peaks. Z peak not evident from SPIKE readings.
![4A](https://user-images.githubusercontent.com/67207661/106677930-9ba79500-65b1-11eb-87fe-8bfa597813c9.png)
![4B](https://user-images.githubusercontent.com/67207661/106677935-9d715880-65b1-11eb-9752-bce183c8ee00.png)


**5)** Good Y and X peaks, but apparently not consistent with SPIKE readings.
![5A](https://user-images.githubusercontent.com/67207661/106677942-9fd3b280-65b1-11eb-97f6-2fd9bbb62026.png)
![5B](https://user-images.githubusercontent.com/67207661/106677944-a19d7600-65b1-11eb-981a-5b9a190b2f31.png)


**6)** Z axis gives a good hint about the collision, but not 100% confirmed at Y neither X.
![6A](https://user-images.githubusercontent.com/67207661/106677956-a530fd00-65b1-11eb-8c8a-ba012c68e429.png)
![6B](https://user-images.githubusercontent.com/67207661/106677961-a6fac080-65b1-11eb-9508-ddaed38840aa.png)


**7)** Hard to infere when the collision occurred, at least through automation.
![7A](https://user-images.githubusercontent.com/67207661/106677970-ab26de00-65b1-11eb-9994-dfc828604b47.png)
![7B](https://user-images.githubusercontent.com/67207661/106677975-ac580b00-65b1-11eb-80d1-20e65cf9663a.png)


**8)** Loads of noise on the Technic hub... Hard to be sure where the impact did happen! All fine with SPIKE hub data.
![8A](https://user-images.githubusercontent.com/67207661/106677980-af52fb80-65b1-11eb-8826-bdfba3597dd0.png)
![8B](https://user-images.githubusercontent.com/67207661/106677984-b11cbf00-65b1-11eb-8a26-136512b89987.png)


Still needing to better look at this data in order to implement some heuristic that fulfills the objectives. And looking forward to a wheel slow down implementation as a complement. 👌

**Reply by laurensvalk** [reply]
Thanks for the detailed analysis! I’ll have to come back to this later when we are having a closer look at this sensor.

While your memory is still fresh, could you add a few notes below each screenshot to say when (approximately) the impact happened and/or how many collisions there were?

**Reply by TechnicBRICKs** [reply]
I've update the post with the charts.

The vehicle was heading towards a wall in each test run, hence there is only one collision per chart.
Both Technic and SPIKE hubs logging were initiated and stopped manually, through Pybricks and SPIKE app, hence these were not synch, neither the exact moment of the collision is registered. I've positioned the collision lines at my best guess for each run.
There were already some notes for each run before the charts, which I've now updated in some cases.

Hope this will turn useful.

**Comment by ZPhilo** [comment]
I had another idea, which was to monitor current consumption using hub.battery.current(). I got it easily working using unregulated motor  but current with regulated motor is much more shaky and would need better filtering to get it working. In the charts attached, yellow is the unfiltered current, pink is shortly damped current, blue is heavily damped one.
Unregulated mode chart
![unregulated](https://user-images.githubusercontent.com/5137285/106785276-841dea00-664d-11eb-98eb-f24875ad5b0b.jpg)
Regulated mode chart
![regulated](https://user-images.githubusercontent.com/5137285/106785273-83855380-664d-11eb-8f87-0a0b0df45c8f.jpg)
Code used (don't remember how to show a .py code in a readable form, "code" doesn't seem to work! - hence the zip)
[currentmonitor-dcmode.zip](https://github.com/pybricks/support/files/5920383/currentmonitor-dcmode.zip)

**Reply by dlech** [reply]
> don't remember how to show a .py code in a readable form

Like this:

    ```python
    # your code here
    ```

**Reply by dlech** [reply]
Extracted from zip.

```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor, UltrasonicSensor
from pybricks.parameters import Port, Direction
from pybricks.tools import wait, StopWatch

# Initialize motors.
left = Motor(Port.A)
right = Motor(Port.B, Direction.COUNTERCLOCKWISE)

# Initialize hub and color sensor.
# You can also use the TechnicHub and/or the ColorDistanceSensor instead.
hub = InventorHub()
eyes = UltrasonicSensor(Port.D)
eyes.lights.on(100)

watch=StopWatch()

longdamp=10
shortdamp=3

#Go!
left.dc(100)
right.dc(100)
wait(100)

avgcurlong=hub.battery.current()
avgcurshort=hub.battery.current()
while watch.time()<5000:
    avgcurlong=(avgcurlong*longdamp+hub.battery.current())/(longdamp+1)
    avgcurshort=(avgcurshort*shortdamp+hub.battery.current())/(shortdamp+1)
    print(avgcurlong, ";", avgcurshort, ";", hub.battery.current())
    wait (20)
```

**Reply by TechnicBRICKs** [reply]
This could be a good idea indeed!

However at least in my case, I want to use a relative lightweight car and small wheels.
The wheels will easily slip on the floor when the car is blocked. Even if the motors will slightly slow down, I've serious doubts there will be any sensible effect in the motors current consumption.

**Comment by laurensvalk** [comment]
Separately, we've discussed detecting collision by looking at the motors/wheels.

I'm starting to look at exposing the approximation of the load/effort in the user API. In essence, this tells you how much feedback torque the motor delivers. It is kind of similar to `stalled`, but much more refined than simply `True` or `False`.

https://user-images.githubusercontent.com/12326241/107499568-62fd5200-6b95-11eb-8bfe-88fba5acdba9.mov

It seems to work quite well in this test. (There's nothing particularly special about this vehicle design; it's just the segway I built earlier with some support wheels. But it does not need to be flat.)

Like acceleration, the raw load estimate is a bit noisy, but in this example I've filtered it a bit, giving a tradeoff between accuracy and response time.

Using it would be fairly straight forward:


```python
load = my_motor.control.load()

if load > threshold_value:
    do_something()
```

I think I'll let this method return values in millinewtonmeters. You could then use @ZPhilo's motor page to get some ideas for thresholds.

Ideas? Thoughts?

**Reply by laurensvalk** [reply]
As a random thought, you could also make a torqe/force sensor this way...

**Reply by TechnicBRICKs** [reply]
It looks great! 
I see you use much larger wheels than I can for my project, which should provide a lot of grip.
I might give it a try over the weekend, with my wheels.

Can't seem to find the @ZPhilo's motor page, to learn about the thresholds.
Can you also share your filtering method code? Some moving average!?

**Reply by laurensvalk** [reply]
Yes, this may still not be perfect for your particular project, but it may be a fairly useful in general.

Philo's motor page is [quite a classic](https://www.philohome.com/motors/motorcomp.htm). You can see e.g. the maximum torque for each motor.

> Can you also share your filtering method code? Some moving average!?

It would be built-in, so that simply reading `load` would give a useful value without additional filtering. Yes, probably a moving average of sorts.

**Reply by TechnicBRICKs** [reply]
Sure!
Didn't got you were referring to Philo's page. Thought it was something else here at github... 😜

Can you share the threshold you used in your example, as a unit reference for that motors and comparation with Philo's data?

**Reply by laurensvalk** [reply]
In this example I used 35 mNm, which is roughly a quarter of the maximum. This is therefore much more sensitive than waiting for a complete stall.

You could set it lower, but that way you will sometimes get false positives. But for a bumper car maybe that isn't soo bad.

**Reply by TechnicBRICKs** [reply]
Indeed, not so bad with some false positives!
So that I've also planned for some randomly triggered direction "reversals" in addition to the bump triggered ones.

**Reply by laurensvalk** [reply]
OK, sensitivity is pretty good. Can detect a piece of paper. Watch the hub light.

https://user-images.githubusercontent.com/12326241/107512478-30f4eb80-6ba7-11eb-8ea0-e76edd6d0151.mov

**Reply by ZPhilo** [reply]
Wow!!!

**Reply by TechnicBRICKs** [reply]
Wow, indeed!

**Reply by laurensvalk** [reply]
This is now documented [here](https://docs.pybricks.com/en/latest/motors.html#pybricks._common.Control.load) and will be useable in the next beta, or sooner if you use the nightly builds.

**Reply by TechnicBRICKs** [reply]
The latest post in v3 Beta Announcements refers to beta.5, but Pybricks Code reports beta.6.
So you mean beta.7 right!?

**Reply by laurensvalk** [reply]
You can try the [latest build, which has the load method, using these instructions](https://docs.pybricks.com/en/latest/start_pup.html#saving-a-program-permanently-beta).

There's no beta build yet. We usually do one every couple of weeks. So if you need it now, try the latest build.

**Reply by dlech** [reply]
and just a warning, Bluetooth has issues currently in the CI builds

**Comment by TechnicBRICKs** [comment]
I don't know what you have done regarding the IMU readings, in the past 4,5 years. 
But just wanted to let you know, that finally... I managed to find the time and put my bumper cars circuit working as intended, and it worked this time at first attempt!
The IMU acceleration reading `hub.imu.acceleration(Axis.X)`, have "now" allowed to detect collisions without doubts, or maybe I just got smarter with time... 😁

I've the model currently at an exhibition that I'm not attending, but soon I'll post a video somewhere, for you to see the show.

**Reply by laurensvalk** [reply]
> I don't know what you have done regarding the IMU readings, in the past 4,5 years... it worked this time at first attempt! ... maybe I just got smarter with time... 

... a bit of both :smile:
